### PR TITLE
ADBDEV-4688: Disable row trigger processing in Orca

### PR DESCRIPTION
**Pull request by hilltracer**
Disable row trigger processing in Orca

Support for  triggers is limited due  to  the distributed nature of GPDB. Using
triggers  can lead to issues because of  unclear  support status and  bugs. For
example, `AFTER UPDATE` triggers are not supported in Orca, and using 
`BEFORE UPDATE` triggers with distributed keys can lead to unexpected errors.

Currently, DML triggers have been removed from Orca in GPDB 7. Orca doesn't plan
queries that fire DML triggers. These queries are planned with Postgres planner.

To prevent potential problems, the `optimizer_enable_dml_triggers` GUC has been
disabled. As  a result, row trigger  handling in GPDB 6 and GPDB 7 is consistent.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/77726

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/77790

**Comment by BenderArenadata** [comment]
Failed job Orca unittests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1782216

**Comment by BenderArenadata** [comment]
Failed job Orca unittests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1782217

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1782214

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1782215

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/77834

**Comment by andr-sokolov** [comment]
Fix description as we discussed today

**Comment by andr-sokolov** [comment]
The commit author name is "hilltracer". Write your name and surname

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/78500

**Comment by RekGRpth** [comment]
Could you please make tests for this patch?

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1818850

File: src/test/regress/expected/gporca_optimizer.out [hunk 1]
```diff
@@ -15438,4 +15438,52 @@ WITH e AS (
 (50 rows)
 
 DROP TABLE d, r;
+-- Test disable optimizer_enable_dml_triggers GUC 
+CREATE TABLE d (t1 int, t2 int, t3 int) DISTRIBUTED BY (t1);
+-- The next query should be processed by Orca
+EXPLAIN (ANALYZE on, COSTS off, VERBOSE off) INSERT INTO d VALUES (1, 1, 1);
```
Comment (andr-sokolov) [inline src/test/regress/expected/gporca_optimizer.out]: Why is the query plan meaningful for the test?

File: src/test/regress/sql/gporca.sql [hunk 2]
```diff
@@ -3706,6 +3706,27 @@ WITH e AS (
 
 DROP TABLE d, r;
 
+-- Test disable optimizer_enable_dml_triggers GUC 
+CREATE TABLE d (t1 int, t2 int, t3 int) DISTRIBUTED BY (t1);
```
Comment (andr-sokolov) [inline src/test/regress/sql/gporca.sql]: Can we remove t3? What does mean "t"? What about c1, c2.. (column)?

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/78637

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/78660

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1827036

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1827034

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/78681

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1827799

**Review by RekGRpth — DISMISSED** [review]

File: src/test/regress/sql/gporca.sql [hunk 3]
```diff
@@ -3706,6 +3706,27 @@ WITH e AS (
 
 DROP TABLE d, r;
 
+-- Test disable optimizer_enable_dml_triggers GUC
```
Comment (andr-sokolov) [inline src/test/regress/sql/gporca.sql]: "Check that optimizer_enable_dml_triggers is defunct"

File: src/test/regress/sql/gporca.sql [hunk 4]
```diff
@@ -3706,6 +3706,27 @@ WITH e AS (
 
 DROP TABLE d, r;
 
+-- Test disable optimizer_enable_dml_triggers GUC 
+CREATE TABLE d (с1 int, с2 int) DISTRIBUTED BY (с1);
+-- The next query should be processed by Orca
+INSERT INTO d VALUES (1, 1);
+CREATE OR REPLACE FUNCTION trig_proc() RETURNS TRIGGER
```
Comment (andr-sokolov) [inline src/test/regress/sql/gporca.sql:3713]: Add empty line before and after the function

File: src/test/regress/sql/gporca.sql [hunk 5]
```diff
@@ -3706,6 +3706,27 @@ WITH e AS (
...
+  RAISE NOTICE 'Values: (%, %)', OLD.с1, OLD.с2;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+CREATE TRIGGER d_upd AFTER UPDATE ON d
+FOR EACH ROW EXECUTE PROCEDURE trig_proc();
+
+SET optimizer_enable_dml_triggers = TRUE;
+-- The next query should be processed by the Postgres planner
+-- because trigger execution in Orca was disabled.
```
Comment (andr-sokolov) [inline src/test/regress/sql/gporca.sql]: was -> has been
execution -> planning

File: src/test/regress/sql/gporca.sql [hunk 6]
```diff
@@ -3706,6 +3706,27 @@ WITH e AS (
...
+UPDATE d SET с2 = 2 WHERE с1 = 1;
+RESET optimizer_enable_dml_triggers;
```
Comment (andr-sokolov) [inline src/test/regress/sql/gporca.sql]: what about replacing RESET with comment why we don't need RESET?

**Comment by andr-sokolov** [comment]
User can not modify the optimizer_enable_dml_triggers variable. So let's remove the variable usage

File: src/test/regress/sql/gporca.sql [hunk 7]
```diff
@@ -3706,6 +3706,33 @@ WITH e AS (
...
+  RAISE NOTICE 'Values: (%, %)', OLD.с1, OLD.с2;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER d_upd AFTER UPDATE ON d
+FOR EACH ROW EXECUTE PROCEDURE trig_proc();
+
+-- GUC should be disabled by default and cannot be enabled
+RESET optimizer_enable_dml_triggers;
```
Comment (andr-sokolov) [inline src/test/regress/sql/gporca.sql]: We assume that all GUSs has default values before tests, so this RESET is redundant

File: src/test/regress/sql/gporca.sql [hunk 8]
```diff
@@ -3706,6 +3706,31 @@ WITH e AS (
...
+
+-- The GUC should be disabled by default and cannot be enabled
+SHOW optimizer_enable_dml_triggers;
+SET optimizer_enable_dml_triggers = TRUE;
+SHOW optimizer_enable_dml_triggers;
+
+-- The next query should be processed by the Postgres planner
+-- because trigger planning in Orca has been disabled.
+UPDATE d SET с2 = 2 WHERE с1 = 1;
+DROP TABLE d;
```
Comment (andr-sokolov) [inline src/test/regress/sql/gporca.sql:3733]: Drop the function too

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/78791

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1832456

**Review by RekGRpth — APPROVED** [review]

**Review by andr-sokolov — APPROVED** [review]

**Pull request merged**